enhancements/network/on-prem-service-load-balancers.md

```diff
@@ -229,35 +230,119 @@ doing all of your testing in a virtual network built on top of it.
-additional component to be installed on on-premise clusters.  However, in the
-[ROADMAP.md
+The MetalLB operator will be managed by OLM as an optional additional component
+to be installed on on-premise clusters.  However, in the [ROADMAP.md
 document](https://github.com/openshift/enhancements/blob/master/ROADMAP.md),
 there is an item to "Front the API servers and other master services with
 service load balancers".  If this functionality is required at install time,
-the details on management of this operator may be revisited.
+the details on management of this operator may be revisited, since MetalLB may
+become a hard requirement in some environments.
 
 There is a start of a
 [metallb-operator](https://github.com/cybertron/metallb-operator) available and
 a [video demo](https://www.youtube.com/watch?v=WgOZno0D7nw).
 
-#### Alternative Integration: Cloud Controller Manager
-
-An alternative integration approach would be via a cloud controller manager
-(CCM). An example of this is the [packet.net
-CCM](https://github.com/packethost/packet-ccm), which ensures MetalLB is
-deployed and also configures it properly to work in packet.net’s BGP
-environment.
-
-These integration options must be explored in more detail as part of a more
-detailed integration proposal.
+#### Configuration Interfaces
+
+MetalLB's current configuration interface is a single `ConfigMap`.  We only
+want to support CRD based configuration through OpenShift.  The ultimate goal
+will be to enhance MetalLB with a native CRD based configuration interface.
+The design work has begun here:
+
+https://github.com/metallb/metallb/pull/833
+
+Since this is an upstream effort, the completion of this work may take longer
+than the timeline for when we want to ship the first piece of MetalLB support
+for OpenShift, which will be to support MetalLB in layer2 mode only.
+
+To mitigate this timing risk, we have another option for a way to configure the
+layer2 mode of MetalLB by adding a small new feature to OpenShift's existing
+network config object.  OpenShift has an existing feature to automatically
+assign external IPs to services of `type=LoadBalancer`.  For example, take this
+sample configuration:
+
+```yaml
+apiVersion: config.openshift.io/v1
+kind: Network
+metadata:
+  name: cluster
+spec:
+  ...
+  externalIP:
+    autoAssignCIDRs:
+    - 10.10.10.0/24
+```
+
+This makes OpenShift automatically set the `externalIP` field, but nothing will
+make sure those addresses are reachable automatically.  What we need is a
+similar option that instead of setting `ExternalIPs`, it should set the
+`LoadBalancerIP` field in the Service spec.  The proposed new field is
+`loadBalancer.autoAssignCIDRs`.  With the new option, the way to configure the
+address pool for MetalLB would be:
+
+```yaml
+apiVersion: config.openshift.io/v1
+kind: Network
+metadata:
+  name: cluster
+spec:
+  ...
+  loadBalancer:
+    autoAssignCIDRs:
+    - 10.10.10.0/24
```

Comment:
I think the naming should make it clearer that this is only expected to be useful in combination with MetalLB.

Additionally, given that fact, there's no reason this assignment necessarily needs to be configured in the cluster network config, or be implemented by openshift-controller-manager. It seems like this is something that would be useful to other MetalLB users too, and could perhaps be done upstream?

Also also, if it does end up in `Network.config.openshift.io`, you should not assume that the external IP config is necessarily a good model to follow in terms of how the config should look. That had the additional constraint that it had to be compatible with what we had done in 3.x.

Reply:
> I think the naming should make it clearer that this is only expected to be useful in combination with MetalLB.

Would you say that the spec.LoadBalancerIP field for Services is also only useful for metallb?  I don't know the history of that field, or what other cloud providers can honor it, if any.

From an OpenShift perspective, indeed it's only metallb right now, but I expected it *could* be useful for some other cloud if any other supported that field, but I haven't gone looking for what else supports it.

> Additionally, given that fact, there's no reason this assignment necessarily needs to be configured in the cluster network config, or be implemented by openshift-controller-manager. It seems like this is something that would be useful to other MetalLB users too, and could perhaps be done upstream?

I answered this a bit to your other comment, but MetalLB does have its own IPAM, but no CRD config interface.

The goal here if we zoom out a bit is supporting a CRD based interface for configuring metallb in its layer2 mode.  This proposal actually started because we thought the existing externalIP allocator may be good enough, but it turned out metallb ignores that field, so then we proposed creating a new feature that mirrored the existing one.

An alternative would be to do this entirely within the downstream MetalLB operator.  We could have an OpenShift specific MetalLB config CRD, and start by only implementing as little as possible for layer2 mode, but still pursue a more complete CRD interface upstream that supports all features, including BGP configuration.

> Also also, if it does end up in `Network.config.openshift.io`, you should not assume that the external IP config is necessarily a good model to follow in terms of how the config should look. That had the additional constraint that it had to be compatible with what we had done in 3.x.

Fair point - open to suggestions.  It's a pretty simple thing, just need 1 or more CIDRs.